{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Introduction to Pandas for Feature Engineering\n",
    "Welcome! If you are looking for a short and sweet, but effective, introduction to some of the coding steps required to perform feature engineering and data analysis for machine learning, look no further. \n",
    "\n",
    "This notebook will use a genomics data set, analyze it, and prepare it for machine learning modeling. \n",
    "\n",
    "I'm assuming you are familiar with writing and reading code, and that you've touched some Python syntax. I'll help ramp you up, very quickly."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get libraries\n",
    "First, we'll want to reference the open source libraries that make all of this possible. I'm using Pandas as my backbone here, but we'll build ontop of that as we need more and more libraries."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd \n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import multiprocessing\n",
    "from multiprocessing import Pool\n",
    "\n",
    "\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get some data\n",
    "Ok! We're going to pull some genomics data, in particular with respect to its relationshp to human health. This is a public dataset coming from ClinVar.\n",
    "- https://www.ncbi.nlm.nih.gov/clinvar/ "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# !wget ftp://ftp.ncbi.nlm.nih.gov/pub/clinvar/vcf_GRCh37/clinvar.vcf.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load into a Pandas DataFrame"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If you're thinking, now what on Earth is a .vcf file, that makes two of us! Fortunately the kind developers who announced this dataset also have a script that converts this into a csv file. That script is hosted on Github, right here: \n",
    "- https://github.com/arvkevi/clinvar-kaggle/blob/master/process_clinvar.py \n",
    "\n",
    "Let's clone that repository, and run this script."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# !git clone https://github.com/arvkevi/clinvar-kaggle.git"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "clinvar-kaggle/process_clinvar.py:50: FutureWarning: set_value is deprecated and will be removed in a future release. Please use .at[] or .iat[] accessors instead\r\n",
      "  cv_df.set_value(cv_df.CLNSIGCONF.notnull(), 'CLASS', 1)\r\n"
     ]
    }
   ],
   "source": [
    "!cp clinvar.vcf.gz /home/ec2-user/SageMaker/clinvar-kaggle\n",
    "!python clinvar-kaggle/process_clinvar.py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "!cp /home/ec2-user/SageMaker/clinvar_conflicting.csv raw_data.csv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If you're able to run that copy command, then you successfully converted the vcf file! Congrats. Now, let's load that into a dataframe. Based on the Python KungFu that was performed by Kevin Arvkei, now we can use the single most import function in pandas `pd.read_csv()`. \n",
    "\n",
    "You'll want to make sure you have `import pandas as pd` in order to use the `read_csv` method. It just needs a csv file name, and it will turn that into a dataframe."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ec2-user/anaconda3/envs/python3/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2785: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    }
   ],
   "source": [
    "df = pd.read_csv('raw_data.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As long as that worked, now you'll perform the second most commonly used function in Pandas. `df.head()`. That gives you a visual confirmation on all of your data. Give it a try!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>AF_ESP</th>\n",
       "      <th>AF_EXAC</th>\n",
       "      <th>AF_TGP</th>\n",
       "      <th>CLNDISDB</th>\n",
       "      <th>CLNDN</th>\n",
       "      <th>CLNHGVS</th>\n",
       "      <th>CLNVC</th>\n",
       "      <th>GENEINFO</th>\n",
       "      <th>MC</th>\n",
       "      <th>ORIGIN</th>\n",
       "      <th>CLNVI</th>\n",
       "      <th>CLNDISDBINCL</th>\n",
       "      <th>CLNDNINCL</th>\n",
       "      <th>CLNSIGINCL</th>\n",
       "      <th>CLASS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>955563</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00114</td>\n",
       "      <td>0.00958</td>\n",
       "      <td>MedGen:C3808739,OMIM:615120|MedGen:CN169374</td>\n",
       "      <td>Myasthenic_syndrome,_congenital,_8|not_specified</td>\n",
       "      <td>NC_000001.10:g.955563G&gt;C</td>\n",
       "      <td>single_nucleotide_variant</td>\n",
       "      <td>AGRN:375790</td>\n",
       "      <td>SO:0001583|missense_variant</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>955597</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.42418</td>\n",
       "      <td>0.28255</td>\n",
       "      <td>MedGen:CN169374</td>\n",
       "      <td>not_specified</td>\n",
       "      <td>NC_000001.10:g.955597G&gt;T</td>\n",
       "      <td>single_nucleotide_variant</td>\n",
       "      <td>AGRN:375790</td>\n",
       "      <td>SO:0001819|synonymous_variant</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>955619</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.03475</td>\n",
       "      <td>0.00879</td>\n",
       "      <td>MedGen:C3808739,OMIM:615120|MedGen:CN169374|Me...</td>\n",
       "      <td>Myasthenic_syndrome,_congenital,_8|not_specifi...</td>\n",
       "      <td>NC_000001.10:g.955619G&gt;C</td>\n",
       "      <td>single_nucleotide_variant</td>\n",
       "      <td>AGRN:375790</td>\n",
       "      <td>SO:0001583|missense_variant</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>957568</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.01761</td>\n",
       "      <td>0.00493</td>\n",
       "      <td>0.01418</td>\n",
       "      <td>MedGen:CN169374</td>\n",
       "      <td>not_specified</td>\n",
       "      <td>NC_000001.10:g.957568A&gt;G</td>\n",
       "      <td>single_nucleotide_variant</td>\n",
       "      <td>AGRN:375790</td>\n",
       "      <td>SO:0001627|intron_variant</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>957640</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.03175</td>\n",
       "      <td>0.02016</td>\n",
       "      <td>0.03275</td>\n",
       "      <td>MedGen:C3808739,OMIM:615120|MedGen:CN169374</td>\n",
       "      <td>Myasthenic_syndrome,_congenital,_8|not_specified</td>\n",
       "      <td>NC_000001.10:g.957640C&gt;T</td>\n",
       "      <td>single_nucleotide_variant</td>\n",
       "      <td>AGRN:375790</td>\n",
       "      <td>SO:0001819|synonymous_variant</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  CHROM     POS REF ALT   AF_ESP  AF_EXAC   AF_TGP  \\\n",
       "0     1  955563   G   C  0.00000  0.00114  0.00958   \n",
       "1     1  955597   G   T  0.00000  0.42418  0.28255   \n",
       "2     1  955619   G   C  0.00000  0.03475  0.00879   \n",
       "3     1  957568   A   G  0.01761  0.00493  0.01418   \n",
       "4     1  957640   C   T  0.03175  0.02016  0.03275   \n",
       "\n",
       "                                            CLNDISDB  \\\n",
       "0        MedGen:C3808739,OMIM:615120|MedGen:CN169374   \n",
       "1                                    MedGen:CN169374   \n",
       "2  MedGen:C3808739,OMIM:615120|MedGen:CN169374|Me...   \n",
       "3                                    MedGen:CN169374   \n",
       "4        MedGen:C3808739,OMIM:615120|MedGen:CN169374   \n",
       "\n",
       "                                               CLNDN  \\\n",
       "0   Myasthenic_syndrome,_congenital,_8|not_specified   \n",
       "1                                      not_specified   \n",
       "2  Myasthenic_syndrome,_congenital,_8|not_specifi...   \n",
       "3                                      not_specified   \n",
       "4   Myasthenic_syndrome,_congenital,_8|not_specified   \n",
       "\n",
       "                    CLNHGVS                      CLNVC     GENEINFO  \\\n",
       "0  NC_000001.10:g.955563G>C  single_nucleotide_variant  AGRN:375790   \n",
       "1  NC_000001.10:g.955597G>T  single_nucleotide_variant  AGRN:375790   \n",
       "2  NC_000001.10:g.955619G>C  single_nucleotide_variant  AGRN:375790   \n",
       "3  NC_000001.10:g.957568A>G  single_nucleotide_variant  AGRN:375790   \n",
       "4  NC_000001.10:g.957640C>T  single_nucleotide_variant  AGRN:375790   \n",
       "\n",
       "                              MC  ORIGIN CLNVI CLNDISDBINCL CLNDNINCL  \\\n",
       "0    SO:0001583|missense_variant     1.0   NaN          NaN       NaN   \n",
       "1  SO:0001819|synonymous_variant     1.0   NaN          NaN       NaN   \n",
       "2    SO:0001583|missense_variant     1.0   NaN          NaN       NaN   \n",
       "3      SO:0001627|intron_variant     1.0   NaN          NaN       NaN   \n",
       "4  SO:0001819|synonymous_variant     1.0   NaN          NaN       NaN   \n",
       "\n",
       "  CLNSIGINCL  CLASS  \n",
       "0        NaN      0  \n",
       "1        NaN      0  \n",
       "2        NaN      1  \n",
       "3        NaN      0  \n",
       "4        NaN      0  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(93529, 19)"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "That last command, `df.shape` tell us exactly how many rows we have in this dataset, and how many columns. This is useful information! We'll rely on that again and again."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### For-Loops and missing value detection in Python\n",
    "If you are new to Python, a very common method of developing a solution is called a for-loop. Basically you'll need to create a list of something, but then you get to walk through that list and perform some type of computation. Let's check it out!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CHROM\n",
      "POS\n",
      "REF\n",
      "ALT\n",
      "AF_ESP\n",
      "AF_EXAC\n",
      "AF_TGP\n",
      "CLNDISDB\n",
      "CLNDN\n",
      "CLNHGVS\n",
      "CLNVC\n",
      "GENEINFO\n",
      "MC\n",
      "ORIGIN\n",
      "CLNVI\n",
      "CLNDISDBINCL\n",
      "CLNDNINCL\n",
      "CLNSIGINCL\n",
      "CLASS\n"
     ]
    }
   ],
   "source": [
    "for column in list(df):\n",
    "    print (column)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The line `list(df)` returns a list of the columns in your dataframe. Based on `df.shape`, it should have 19 items. Then we call the `print()` function on each item in the list. Not too bad!\n",
    "\n",
    "Let's see if we can do a bit more."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found 0 missing values for column CHROM\n",
      "Found 0 missing values for column POS\n",
      "Found 0 missing values for column REF\n",
      "Found 0 missing values for column ALT\n",
      "Found 0 missing values for column AF_ESP\n",
      "Found 0 missing values for column AF_EXAC\n",
      "Found 0 missing values for column AF_TGP\n",
      "Found 0 missing values for column CLNDISDB\n",
      "Found 0 missing values for column CLNDN\n",
      "Found 0 missing values for column CLNHGVS\n",
      "Found 0 missing values for column CLNVC\n",
      "Found 6948 missing values for column GENEINFO\n",
      "Found 8577 missing values for column MC\n",
      "Found 6948 missing values for column ORIGIN\n",
      "Found 59697 missing values for column CLNVI\n",
      "Found 93402 missing values for column CLNDISDBINCL\n",
      "Found 93402 missing values for column CLNDNINCL\n",
      "Found 93402 missing values for column CLNSIGINCL\n",
      "Found 0 missing values for column CLASS\n"
     ]
    }
   ],
   "source": [
    "for column in list(df):\n",
    "    \n",
    "    missing_values = df[column].isna().sum()\n",
    "    \n",
    "    print ('Found {} missing values for column {}'.format(missing_values, column))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Looks like some of those columns have very high frequencies for missing values! Let's dissect it even further, and figure out the percentage that is missing from those columns. We'll use that to figure out if we should just drop those columns, or impute the missing values."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7.0% of values are missing for colum GENEINFO\n",
      "9.0% of values are missing for colum MC\n",
      "7.0% of values are missing for colum ORIGIN\n",
      "64.0% of values are missing for colum CLNVI\n",
      "100.0% of values are missing for colum CLNDISDBINCL\n",
      "100.0% of values are missing for colum CLNDNINCL\n",
      "100.0% of values are missing for colum CLNSIGINCL\n"
     ]
    }
   ],
   "source": [
    "cols_to_check = ['GENEINFO', 'MC', 'ORIGIN', 'CLNVI', 'CLNDISDBINCL', 'CLNDNINCL', 'CLNSIGINCL']\n",
    "\n",
    "for col in cols_to_check:\n",
    "    missing_values = df[col].isna().sum()\n",
    "    percentage = np.round((missing_values/df.shape[0])*100)\n",
    "    print ('{}% of values are missing for colum {}'.format(percentage, col))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Ok! Based on this, we should clearly drop the last 4 columns. Let's do that below, but wrap it as a function so we can store the code and reuse it later. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Functions in Python\n",
    "Functions are a crucial step in developing software. Functions help us keep our code modular, clean, reusable, and user-friendly! They're easier to read, understand, and debug. \n",
    "\n",
    "Functions are easy in Python. They will look like this 100% of the time `def function_name(arguments):`. Then, the next lines HAVE TO either have a single indent, or 4 spaces. If you don't like it, write your own software language."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "def drop_missing_columns(df):\n",
    "    cols_to_drop = ['CLNVI', 'CLNDISDBINCL', 'CLNDNINCL', 'CLNSIGINCL']\n",
    "    new_df = df.drop(cols_to_drop, axis=1)\n",
    "    return new_df\n",
    "\n",
    "df = drop_missing_colums(df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "OK! Now, if you try running that cell twice, the 2nd time you run it will break. The reason? It's actually dropping those columns from your dataframe, so it's going to break if you try to drop it without having anything there. To get around that, let's add another function that reads in your data from disk, so you can easily get a new dataframe for every new step in your ETL flow."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ec2-user/anaconda3/envs/python3/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2903: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  if self.run_code(code, result):\n"
     ]
    }
   ],
   "source": [
    "def read_data():\n",
    "    df = pd.read_csv('raw_data.csv')\n",
    "    return df\n",
    "\n",
    "df = read_data()\n",
    "df = drop_missing_columns(df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Generic Analysis and Plotting in Pandas\n",
    "Ok! You've made it this far. Good job. Now we're going to learn some basic analysis techniques. No matter where you go, as long as you have Python and a csv file, you can use these techniques. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    71492\n",
       "1    22037\n",
       "Name: CLASS, dtype: int64"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['CLASS'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "That's a very common method, `.value_counts()`. You need a Pandas Series, or a column. It's a subset of a dataframe, and one of the reasons why they're so popular! Nice and easy to use. Here we're 'indexing' into the dataframe by using `df['CLASS']` to refence just the column named 'CLASS'. Then, we're using `.value_counts()` to figure out how many of one type of value is present in that column, along with the other. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Class balance is 24.0%\n"
     ]
    }
   ],
   "source": [
    "ratio = 22037 / (22037+71492)\n",
    "print ('Class balance is {}%'.format(np.round(ratio * 100)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This is useful information! This is telling us that out of our entire dataset, 24% of the cases fall into the positive category, into conflicting diagnoses. We are trying to build a model to detect when a diagnosis will be conflicting, so this is extremely relevant."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Distribution Analysis\n",
    "Now, let's loop through all of our remaining columns, and print out the distributions we have. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[<matplotlib.axes._subplots.AxesSubplot object at 0x7fe953c8cfd0>,\n",
       "        <matplotlib.axes._subplots.AxesSubplot object at 0x7fe954afa358>],\n",
       "       [<matplotlib.axes._subplots.AxesSubplot object at 0x7fe951ddf8d0>,\n",
       "        <matplotlib.axes._subplots.AxesSubplot object at 0x7fe954200e48>],\n",
       "       [<matplotlib.axes._subplots.AxesSubplot object at 0x7fe954212400>,\n",
       "        <matplotlib.axes._subplots.AxesSubplot object at 0x7fe953b33978>]],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYcAAAEVCAYAAAALsCk2AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvnQurowAAIABJREFUeJzt3XucHGWd7/HPF8LNcAkQnWUJMrrm7DHCcouAgmeDCAZYCSoiiCZwOGTPAbyBl7iLC6IeAUVZENEgMaAgIILEAwhZYATPEpZEsiTcliwbJOESciEQYIXAb/+oZ0zR1TPTPTPdXd3zfb9e/Zqup56qfqry6/y6nqp6ShGBmZlZ3katboCZmZWPk4OZmRU4OZiZWYGTg5mZFTg5mJlZgZODmZkVODmYmVmBk4OZmRU4OZSMpB5JayRtliubLekVSetyr48PsJ6lkl6uWOb7ad6mks6TtCyVL5V0fh/LPpM+f8vGbbWNFE2K709LWixp01z9z0m6T9KoXNmWabmb+/iMT0ian+o8JelmSfsPfS+0ByeHEpHUDbwPCODwitnnRsSWudfVNazyQxXLnJLKvwJMBPYGtgImAb+vtiywZ6p7+mC2yaxXE+P7IuA54O/T574d+BpwQkSszy3/UeCPwEGS/qyiracC5wP/F+gC3gr8AJhS8wa3OSeHcpkKzANmA9Ma+DnvBq6PiCcjszQiLq9WMSKWAzcDuzSwPTYyNCW+I+J14ATg85J2BS4BfhARlT+ApgE/BO4HPtlbKGkb4Czg5Ii4LiJejIhXI+LXEfHFRrW7bEYNXMWaaCrwXeAeYJ6kroh4pgGfMw84VdIrwF3A4uhjkC1JOwGHAtc1oB02sjQrvomIRyR9C7gDWAUclp8vaWeyI+ZTgNVkieI7afZ7gM2B6xvRtnbhI4eSSH2ZOwPXRMQC4N+BT+SqfEHSc+m1ssbV/iq3zHOSTkzl3wLOAY4F5gPLJVX+kvuVpOeA3wG/JTu8NhuUJsd3r7uA7YFrI+I/K+Z9Crg/Ih4ErgLeJWmPNG97YGVFF9SI4+RQHtOAWyOi94txJW889P5ORIxJr7E1rvOI3DJjIuISgIh4LSIuioj9gDHAN4FZkt5ZZdmdI+KkiHh5iNtnI1vT4huyiy6AHwEXAqek8w55U4Er4E9dp7/NtWcVMDZ/8nokGtEbXxaStgCOAjaW9HQq3gwYI2m3Rn52+k//IklfAyYADzXy82zkaVF8fxVYAXwWeJksURyU2vNeYDzwFUmnpfpbAbtI+gJwN9mJ6iOAaxvUvtJzciiHI4DXgF2BV3Ll15D9whlWkj4HLCTr+32VrHtpK+C+4f4sM5of37sBnwH2ioiQdCZwv6TjI+InZEcIcys+ewuyE9OHRMSvJf0D2Y+m9cCtZN+TDwAHRMSXhrvNZeTkUA7TgJ9ExB/yhem67QuAfxrken8t6bXc9NyI+DDwEnAe8A6yywr/DfhoRDw2yM8x60/T4hs4ErgU+GZELIHs6Didj7hW0u1kRzFTI+Lp/Mok/TS19dcRcV46yjmdrPvpBWABWRfsiCA/Cc7MzCr5hLSZmRW4W6lNSXor8GAfsydUHsKbtRPHd+u5W8nMzAra9shh7Nix0d3dXSh/8cUXGT16dPMbVDLeDxv0tS8WLFiwMiLe3IImDYpjfmDeF5nhiPm2TQ7d3d3Mnz+/UN7T08OkSZOa36CS8X7YoK99Ienx5rdm8BzzA/O+yAxHzPuEtJmZFbTtkUNfFi1fy3EzbqxrmaVnHzZwJbOSGkzMg+Pe+ucjBzMzK3ByMDOzAicHMzMrcHIwM7MCJwczMytwcjAzswInBzMzK3ByMDOzAicHMzMrcHIwM7MCJwczMytwcjAzswInBzMzK3ByMDOzAicHMzMrcHIwM7MCJwczMytwcjAzswInBzMzK3ByMDOzAicHMzMrqCk5SFoqaZGkhZLmp7LtJM2V9Gj6u20ql6QLJC2RdL+kPXPrmZbqPyppWq58r7T+JWlZDfeGmtXDMW8jXT1HDgdExO4RMTFNzwBui4jxwG1pGuAQYHx6TQcuhuyLBZwB7APsDZzR++VKdU7MLTd50FtkNnwc8zZiDaVbaQpwWXp/GXBErvzyyMwDxkjaAfggMDciVkfEGmAuMDnN2zoi5kVEAJfn1mVWJo55GzFG1VgvgFslBfCjiJgJdEXEU2n+00BXer8j8ERu2WWprL/yZVXKCyRNJ/tlRldXFz09PYU6XVvAabuur3GzMtXW0+7WrVvXkds1GIPcF6WIebNWqTU57B8RyyW9BZgr6eH8zIiI9CVqqPQFnQkwceLEmDRpUqHOhVfcwHmLat2szNJji+tpdz09PVTbPyPRIPdFKWK+UT+IwD+KOtlw7Iea/heNiOXp7wpJ15P1nz4jaYeIeCodJq9I1ZcDO+UWH5fKlgOTKsp7Uvm4KvXNWqYsMd+oH0TgH0WdbDj2w4DnHCSNlrRV73vgYGAxMAfovfpiGnBDej8HmJqu4NgXWJsOxW8BDpa0bTopdzBwS5r3vKR90xUbU3PrMms6x7xZbUcOXcD16Uq7UcCVEfEbSfcC10g6AXgcOCrVvwk4FFgCvAQcDxARqyV9Hbg31TsrIlan9ycBs4EtgJvTy6xVHPM24g2YHCLiMWC3KuWrgAOrlAdwch/rmgXMqlI+H9ilhvaaNZxj3sx3SJuZWRVODmZmVuDkYGZmBU4OZmZW4ORgZmYFTg5mZlbg5GBmZgVODmZmVuDkYGZmBU4OZmZW4ORgZmYFTg5mZlbg5GBmZgVODmZmVuDkYGZmBU4OZmZW4ORgZmYFTg5mZlbg5GBmZgVODmZmVuDkYGZmBU4OZmZW4ORgZmYFTg5mZlZQmuQgabKkRyQtkTSj1e0xazTHvJVZKZKDpI2Bi4BDgAnAMZImtLZVZo3jmLeyG9XqBiR7A0si4jEASVcBU4AHm/Hh3TNubMbHNNXsyaNb3QTrX0tjHhz31r+yJIcdgSdy08uAfSorSZoOTE+T6yQ9UmVdY4GVw97CNnPAOd4POX3ti52b3ZAcx3wDOO7/ZMgxX5bkUJOImAnM7K+OpPkRMbFJTSot74cN2nlfOObr432RGY79UIpzDsByYKfc9LhUZtapHPNWamVJDvcC4yW9TdKmwNHAnBa3yayRHPNWaqVIDhGxHjgFuAV4CLgmIh4Y5Or6PQQfQbwfNijdvnDMF0n6hKT5ktZJekrSzZL2l3SmpJ8NsGyPpDXArIrycZJ+KWmlpLWSFks6Ljf/BEkPS3pB0jOSbpK0VWO2sKmGHhMR4VdJXkAPsAbYLFc2G3gFWJd7fbyfdeTrvQ68nJs+NtUZD1wFPAs8DzwKXAiMS/MnpWXXAS8AjwDHt3r/+NW5L+BUYAXwEWA0sAnwIeDbwJnAz/pZtht4DVgNfKxi3h3A+Wmdo4A9gEPSvL8GngH2SNPbAdOArVq9P8rwKsWRg4GkbuB9QACHV8w+NyK2zL2u7ms9+XrAH4AP5cqukPQO4B7gSbIvxdbAfsC/A/vnVvVkWsfWwJeBS3wdvjWCpG2As4CTI+K6iHgxIl6NiF9HxBdrWMVUYB7ZD6lpFfPeDcxO61wfEfdFxM25eXdHxH0AEbE6Ii6LiBeGZcPanJNDefQX4MPpTOD/R8SpEbEMICJWRMT5EXFVZeXI/IrsiMbJwRrhPcDmwPWDXH4qcEV6fVBSV27ePOAiSUdLemvFcvek+l+TtJ+kzQb5+R2pbZPDQEMPSNpM0tVp/j3pl3mZ9RfgfaphCIaDJT0raaGkhWSH6r+stVGSNpL0YWAMsKjW5ZpN0ixJKyQt7mO+JF2Q9tP9kvZsdhuHqgNjvtf2wMrIzsMMqGI/XEx27f41EbGA7Aj4e70xD3QBa4GvAv+RvgfvBoiIu8i6sfYEbgRWSfpuunu99Boe863u1xrMC9iYLAjeDmwK/CswoaLOScAP0/ujgatb3e5+tmd/4FVgbJp+GPh8ej8b+E/gufRaWet+AJYC5wLfz5WtBybnpk9J610HXJLKJpGdc3iOrB93IXB0q/fTAPvwf5B9yRf3Mf9Q4GZAwL7APa1uc53b11ExX9HuySkuR/Ux/0zSOYcq+2EV8Ntc3X8g6079fpX1jE3fp+WAKuZtBByY4v1vW71PatxvDY35dj1y+NPQAxHxCtnJ1SkVdaYAl6X31wIHSlIT21iPacCtEdF7R+OVvLFr6TsRMSa9xubKa9kPlVYBO/RORMT3I2IM2Um7TXL1nkyft11E7B5VupzKJCLuJPti92UKcHlk5gFjJO3QT/2y6bSYz7sb+CNwRA1188OObEx2onlfSU9Lehr4PNn9I9tXLpi+X98B/pzs5HN+3usRcRtwO7DLELalaRod8+2aHKoNPbBjX3UiO1xdS5WAaTVJWwBHAX9dEeC7SdptgMVr2Q8AH02HldeSfRE/MgxNbze17quy6piYrxQRa8l+8V8k6QhJb5K0iaRDJJ2bqm0kaXOyK5OeTOcHjiA74vg5sHt6vRP4N+BDKeYflvQBSaPSJar/hyy5rJI0JZ2L2DZ1wexNdgXTvGZufwMNKebbNTl0kiPILsObwBsD/C6y8xBDdTfQHRF/Bcwl64N9X+pb3RFA0tj0mWYtERHnkV3OejrZJdZPkHV5/ipVOYbssuwrgePIupamAXcC6yLi6d4XcA5ZN+meZFflzSHrIn2M7PxE79WAa4ATyS7lfh74GfDtiLiikdvaLtpqbKWcWoYe6K2zTNIoYBuyLpWymQb8JCL+kC+U9H3gAuCf+lm2lv3wQkT8Mb3/Mdk5iL2BrwP/mn6BPQncmuZ1qnYfrqKTYr6q9J9ytf+Y/5nsvAOS3gOcGREfTNNfqbKeWaSb4SQdBKyOiG2q1LuT7DxDpxpazLf6pMogT8SMIvsV8DY2nJx7V0Wdk3njyblrWt3uFu2HHXLvPwzMa3W7G7g/uun75NxhvPHk3L+0ur0N+Ld2zIdjPjdvSDHflkcOEbFeUu/QAxsDsyLiAUlnAfMjYg5wKfBTSUvITtoc3boWN0aN++Ezkg4n65tdTXZI3nEk/ZzsKquxkpYBZ5BOsEfED4GbyK7eWAK8BBzfmpYOjmM+45jfoNExr5RhrI2km3n6eijMhKjoojIzq5eTg5mZFbRltxLA2LFjo7u7u1D+4osvMnq0HxXo/bBBX/tiwYIFKyPizS1o0qA45gfmfZEZjphv2+TQ3d3N/PnzC+U9PT1MmjSp+Q0qGe+HDfraF5Ieb35rBs8xPzDvi8xwxLzvczAzs4K2PXLoy6Llazluxo11LbP07MMa1BqzxhtMzIPj3vrnIwczMytwcjAzswInBzMzK3ByMDOzAicHMzMrcHIwM7MCJwczMytwcjAzswInBzMzK3ByMDOzAicHMzMrcHIwM7MCJwczMytwcjCrQtJSSYskLZQ0P5VtJ2mupEfT321TuSRdIGmJpPsl7Zlbz7RU/1FJ03Lle6X1L0nLqvlbadY3Jwezvh0QEbtHxMQ0PQO4LSLGA7elaYBDgPHpNR24GLJkQvbQ932AvYEzehNKqnNibrnJjd8cs9o5OZjVbgpwWXp/GXBErvzyyMwDxkjaAfggMDciVkfEGmAuMDnN2zoi5kX2EPfLc+syK4WOe9iP2TAJ4FZJAfwoImYCXRHxVJr/NNCV3u8IPJFbdlkq6698WZXyAknTyY5G6Orqoqenp1Cnaws4bdf19WwbQNV1tbt169Z15HbVazj2g5ODWXX7R8RySW8B5kp6OD8zIiIljoZKSWkmwMSJE6Pac4EvvOIGzltU/1d56bHFdbU7P0M6Mxz7wd1KZlVExPL0dwVwPdk5g2dSlxDp74pUfTmwU27xcamsv/JxVcrNSsPJwayCpNGStup9DxwMLAbmAL1XHE0Dbkjv5wBT01VL+wJrU/fTLcDBkrZNJ6IPBm5J856XtG+6Smlqbl1mpeBuJbOiLuD6dHXpKODKiPiNpHuBaySdADwOHJXq3wQcCiwBXgKOB4iI1ZK+Dtyb6p0VEavT+5OA2cAWwM3pZVYaNSUHSUuBF4DXgPURMTFdpnc10A0sBY6KiDXpl9A/kn1ZXgKOi4jfp/VMA05Pq/1GRFyWyvdiwxflJuCz6SoOs6aLiMeA3aqUrwIOrFIewMl9rGsWMKtK+XxglyE31qxB6ulW8jXfZmYjxFDOOfiabzOzDlXrOYeOvua7E6+L9vXeG3hfmNWv1uTQ0dd8+3rvzuZ9YVa/mrqVfM23mdnIMmBy8DXfZmYjTy39L77m28xshBkwOfiabzOzkcfDZ5iZWYGTg5mZFTg5mJlZgZODmZkVODmYmVmBk4OZmRU4OZiZWYGTg5mZFTg5mJlZgR8TamZWYt0zbqx7mdmTRw/5c33kYGZmBU4OZmZW4ORgZmYFTg5mZlbg5GBmZgVODmZmVuDkYGZmBU4OZmZW4ORgZmYFTg5mZlbg5GBmZgVODmZmVuDkYGZmBaVJDpImS3pE0hJJM1rdHrNGc8xbmZUiOUjaGLgIOASYABwjaUJrW2XWOI55K7uyPM9hb2BJRDwGIOkqYArwYDM+fDDjpQMsPfuwYW6JjSAtjXmzgZQlOewIPJGbXgbsU1lJ0nRgeppcJ+mRKusaC6wc9hZWoXOa8SmD1rT90Ab62hc7N7shOS2P+ZLH72A57oEDzhl6zJclOdQkImYCM/urI2l+RExsUpNKy/thg3beF475+nhfZIZjP5TinAOwHNgpNz0ulZl1Kse8lVpZksO9wHhJb5O0KXA0MKfFbTJrJMe8lVopkkNErAdOAW4BHgKuiYgHBrm6fg/By07ScZIWSXpJ0tOSLpY0Js07U9KrktZJek7SP0t6T27ZSZKWpcmZqewgSXdIekHSKkkLJX1Z0ua5df4st45In79RruwbkmY3Y/sbpHQx4ZivnaSlkl5Ocf+MpNmStkzz/kbSv0h6UdIq4DVJ43LLbirpPEnL0vJLJZ3fso1pniHHRCmSA0BE3BQR/y0i/iIivjmE9bTtF0XSacA5wBeBbYB9yU4gzU2/LgGujogtyU683QH8otq6ImKmpI8B1wJXAjtHxPbAx8m6MHaqtlzy52S/ZDtCWWPCMV+XD6W43xOYCJwu6Uiy2D6f7PvwLuAB4HeStk3LfSXV3xvYCpgE/L65TW++4YiJ0iSHkU7S1sDXgE9HxG8i4tWIWAocBXQDn8zXT788rwB2lPTmKusT8F3grIi4JCJWp+UeiYhPR8Sj/TTnXOBrktrqggXrfBGxHLgZ2BU4D/hGRFwZES9HxNPA/wLWAZ9Pi7wbuD4inozM0oi4vCWNbzNODuXxXmBz4Lp8YUSsA24CDsqXpyOJqcAqYE2V9f0l2RHCLwfRluuA54HjBrGsWcNI2gk4FHgJeCsVR84R8TpZzPd+X+YBp0o6SdKu6UeT1aBtk8NAQw9I2kzS1Wn+PZK6m9/KuowFVqYjgkpPpfkAR0l6DngZOJHsV/4DkpYAn6hYH8BBkp5N5xrWpHMZL0n6VD9tCeCrwFdz3VmlJmmWpBWSFvcxX5IuSPFwv6Q9m93GoerAmK/Hr1Lc/w54jA33hBxbpW438G5JC4EjgbtSvfnAcknTGt/cxmt0zLdlclBtQw+cAKyJiHcA3yPryy+zlcDYPrpydmDDDS3XRMQYoAtYDHyJDfvh/Wy4d2VV+rsN2XmK3SNi24h4E1mf68b9NSYibiK7MetvB79JTTUbmNzP/EOA8ek1Hbi4CW0aNh0a8/U4IsX928m2/3+m8mOr7IdtgGdTzO8eEcdGxH7AGOCbwCxJ72xayxtnNg2M+bZMDuSGHoiIV4DeoQfypgCXpffXAgeW/JDybuCPwEfyhemqjEOA2/LlEbES+AFZwL+c9sPtwBapyiNk183vNYQ2/T3wd8CbhrCOpoiIO4HV/VSZAlye+p3nAWMk7dCc1g2LToz5wdgbWEL2fVgGLCW3H9JVdhPTvDdI5yUuIuuGbftxrBod8+2aHKoNPbBjX3VSV81aYPumtG4QImIt2QnpC1P3wSapW+Aasu37aZXFXiPbxi+l6WdJRwSp7/U04Ajgk5IekHStpPeRHXXU0qYesqOTTjgMryVmyqzjYn6QdgSeiIgAvkB29dEkSZtL+jPgx2Q/kHZKXSkLJX1M0haSRqUupa2A+1q1AU00pJhv1+TQkSLiXLJf6t8hOyF8D9k/7oER8cc+FrsfmC7pLVXWdzXZSev7yS6JnUx2cnsmfVwCW8XpwHZ1bIZZU6T4/iGwO1k36oNkiWES2aXbfwUsAi4Fnibrmj0Z+GjvgIfWt3a9VLGWoQd66yxL/fjbsKEfvrQi4lKyYK4278yKouXA5hExGkDSOrI+1fwyvyAlgtRvvToivt3XOiNCFdP3AJ3QNdHuw1V0bMwPJCK6c5OV+2EFcH5EfKuPxY8ji/ltGtO6UhtSzLfrkUMtQw/MYUN3yJHA7elQtJMMuB8q+hgPJ7sbdySaA0xNV3DsC6yNiKda3ag6OOYzjvnaDSnm2/LIISLWS+odemBjYFZEPCDpLGB+RMwh+/X903SJ52o66I7fXjXuh89IOhxYT7YfjmtZgxtI0s/JuhPGKhtC5AxgE4CI+CFZd9qhZCczXwKOb01LB8cxn3HMb9DomFfn/bAwM7OhatduJTMza6C27FYCGDt2bHR3dxfKX3zxRUaPHt38BjVIJ21P2bZlwYIFKyOiMC5VWY2UmO/VidvV6m2qJ+bbNjl0d3czf/78QnlPTw+TJk1qfoMapJO2p2zbIunxVrehHiMl5nt14na1epvqiXl3K5mZWUHbHjn0ZdHytRw348a6lll69mENao1ZeXXX+T0Bf1dGEh85mJlZgZODmZkVODmYmVmBk4OZmRU4OZhVkLSTpDskPZiGOv9sKt9O0lxJj6a/26byPp+4JWlaqv9o/glkkvaStCgtc0EHPnfB2pyTg1nReuC0iJgA7AucnJ42NgO4LSLGkz1spvdRnVWfuCVpO7LxbvYhe0jNGb0JJdU5Mbdcf0/0Mms6JwezChHxVET8Pr1/gWxUzx1545PWLiN7kBL0/cStDwJzI2J1RKwB5gKT07ytI2JeGjX18ty6zErBycGsH+lpfHuQPXipKzfk8dNseKJeX0/c6q98WZVys9LouJvgzIZLen73L4HPRcTz+dMCERGSGj6ksaTpZF1VdHV10dPTU6izbt26quUDOW3X9XUvM5jPGazBbleZtdM2OTmYVSFpE7LEcEVEXJeKn5G0Q0Q8lbqGVqTyvp64tZxsvP18eU8qH1elfkFEzCR7rCsTJ06MauPyDHa8nnpHEgBYemz9nzNYrR6HqBHaaZvcrWRWIV05dCnwUER8Nzcr/6S1acANufJqT9y6BThY0rbpRPTBwC1p3vOS9k2fNTW3LrNS8JGDWdF+wKeARZIWprK/A84GrpF0AvA4cFSaV/WJWxGxWtLXyR5tCXBWRKxO708CZgNbADen16AMZjwxs4E4OZhViIjfAX3dd3BglfoBnNzHumYBs6qUzwd2GUIzzRrK3UpmZlbg5GBmZgVODmZmVuDkYGZmBU4OZmZW4ORgZmYFTg5mZlZQU3KQtDSNPb9Q0vxU5rHtzcw6VD1HDgdExO4RMTFNe2x7M7MONZRuJY9tb2bWoWpNDgHcKmlBGkIYPLa9mVnHqnVspf0jYrmktwBzJT2cn1mmse27tqh/nPoyj6/eTuO/D6STtsWs09WUHCJiefq7QtL1ZOcMSjm2/YVX3MB5i+obT7CZY9TXq53Gfx9IJ22LWacbsFtJ0mhJW/W+JxuTfjEe297MrGPV8hO7C7g+XV06CrgyIn4j6V5KOLa9mZkN3YDJISIeA3arUr4Kj21vNqJ0D+bRomcf1oCWWKP5DmmzKiTNkrRC0uJcmW/8tBHDycGsutkUb8b0jZ82YvgxoWZVRMSdkroriqew4Yq7y8iutvsyuRs/gXmSem/8nES68RNAUu+Nnz2kGz9Tee+Nnz7XlnTPuJHTdl1f97Ox3YU1fJwczGrX9Bs/G3VvTzMN5t6W03Zd33H3LEF73evj5GA2CM268bNR9/Y002DuIzouHTl00j1L0F73+vicg1ntnkndRdRx42df5TXd+GnWKk4OZrXzjZ82YpT3WNSshST9nOyE8lhJy8iuOjob3/hZt8HcG2Gt5+RgVkVEHNPHLN/4aSOCu5XMzKzARw5m1jE8vMfw8ZGDmZkVODmYmVmBk4OZmRU4OZiZWYFPSJuZ1WkknPh2cjCzEc036VXnbiUzMytwcjAzswInBzMzK3ByMDOzAicHMzMrcHIwM7MCX8pqZtYE3enRp8c14dLZ4binwkcOZmZWUJrkIGmypEckLZE0o9XtMWs0x7yVWSmSg6SNgYuAQ4AJwDGSJrS2VWaN45i3sitFcgD2BpZExGMR8QpwFTClxW0yayTHvJVaWU5I7wg8kZteBuxTWUnSdGB6mlwn6ZEq6xoLrKznw3VOPbWbru7tKbGybcvOLfzslsZ8O/hMB25Xs7apn//Tao75siSHmkTETGBmf3UkzY+IiU1qUsN10vZ00rY0y0iM+V6duF3ttE1l6VZaDuyUmx6Xysw6lWPeSq0syeFeYLykt0naFDgamNPiNpk1kmPeSq0U3UoRsV7SKcAtwMbArIh4YJCr6/cQvA110vZ00rYMiWO+Jp24XW2zTYqIVrfBzMxKpizdSmZmViJODmZmVtBRyaHdhiOQtJOkOyQ9KOkBSZ9N5dtJmivp0fR321QuSRek7btf0p6t3YIiSRtLuk/S/0vTb5N0T2rz1enkK5I2S9NL0vzuVra7XbVbzNdC0ixJKyQtbnVbhktf3/Uy65jk0KbDEawHTouICcC+wMmpzTOA2yJiPHBbmoZs28an13Tg4uY3eUCfBR7KTZ8DfC8i3gGsAU5I5ScAa1L591I9q0ObxnwtZgOTW92IYdbXd720OiY50IbDEUTEUxHx+/T+BbL/VHcka/dlqdplwBHp/RTg8sjMA8ZI2qHJze6TpHHAYcCP07SA9wPXpiqV29K7jdcCB6b6Vru2i/laRMSdwOpWt2M49fNdL61OSg7VhiMo9c7PS90qewD3AF0R8VSa9TT2u42qAAACq0lEQVTQld6XfRvPB74EvJ6mtweei4j1aTrf3j9tS5q/NtW32pU9HqyKiu96aXVScmhbkrYEfgl8LiKez8+L7Frj0l9vLOlvgBURsaDVbTErq/6+62VTipvghklbDkcgaROyYLkiIq5Lxc9I2iEinkrdRitSeZm3cT/gcEmHApsDWwP/SNb1NSodHeTb27styySNArYBVjW/2W2tzPFgFfr4rpdWJx05tN1wBKmP/VLgoYj4bm7WHGBaej8NuCFXPjVdtbQvsDbX/dRSEfGViBgXEd1k+/72iDgWuAM4MlWr3JbebTwy1S/9EVLJtF3Mj1T9fNdLq2OSQ/pl2jscwUPANUMYjqBZ9gM+Bbxf0sL0OhQ4GzhI0qPAB9I0wE3AY8AS4BLgpBa0uV5fBk6VtITsnMKlqfxSYPtUfiobrsiyGrVpzA9I0s+Bu4G/lLRM0gkDLdMG+vqul5aHzzAzs4KOOXIwM7Ph4+RgZmYFTg5mZlbg5GBmZgVODmZmbaCeAQklvTUN9HdfGqSz7iujnBzMzNrDbGofkPB0skub9yC7/+UH9X6Yk4OZWRuoNiChpL+Q9BtJCyTdJem/91YnG6UAstEHnqz38zpp+Awzs5FmJvC/I+JRSfuQHSG8HzgTuFXSp4HRZDfT1sXJwcysDaVB/N4L/CI32v1m6e8xwOyIOE/Se4CfStolIl6vsqqqnBzMzNrTRmRD4u9eZd4JpPMTEXG3pM2BsWwYxLOmlZuZWZtJQ37/h6SPwZ8eI7xbmv0H4MBU/k6ykZKfrWf9HlvJzKwNpAEJJ5EdATwDnAHcTva44B2ATYCrIuKs9AjSS4AtyU5Ofykibq3r85wczMyskruVzMyswMnBzMwKnBzMzKzAycHMzAqcHMzMrMDJwczMCpwczMys4L8Ansyn3wwHZr4AAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 6 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df.hist()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we're getting somewhere. This is telling us that most of our data is not normally distributed. For `AF_ESP, AF_EXAC, AF_TGP, ORIGIN`, there is a strong burst of data at the begining of the distribuion, but then it dramatically falls off. I'm thinking we should create binary indicators here for the column being in that first group. The `POS` data is still somewhat skewed, but it's nowhere near as stark. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [],
   "source": [
    "def convert_cols_to_binary(df):\n",
    "\n",
    "    cols_to_convert_to_binary = {'AF_ESP': 4, 'AF_EXAC':5, 'AF_TGP':6}\n",
    "\n",
    "    # add new columns \n",
    "    for c in cols_to_convert_to_binary.keys():\n",
    "        df['{}_is_zero'.format(c)] =  0\n",
    "        \n",
    "    for idx, row in df.iterrows():\n",
    "    \n",
    "        for k, v in cols_to_convert_to_binary.items():\n",
    "\n",
    "            # if that value is zero\n",
    "            if row[k] in [0, 0.0]:\n",
    "\n",
    "                # add a binary indicator for that column \n",
    "                df.loc[idx, k] = 1\n",
    "                \n",
    "    return df \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for idx, row in df.iterrows():\n",
    "    \n",
    "    for k, v in cols_to_convert_to_binary.items():\n",
    "\n",
    "        # if that value is zero\n",
    "        if row[k] in [0, 0.0]:\n",
    "\n",
    "            # add a binary indicator for that column \n",
    "            df.loc[idx, k] = 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# dictionaries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# missing values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# one hot encoding"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# detecting anomalies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# unit tests in Python"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# map reduce for feature engineering "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# correlation & decorrelation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# package your ETL code as an inference pipeline "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# splitting "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "conda_python3",
   "language": "python",
   "name": "conda_python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
